File: src/main/java/work/nekow/particledrawing/core/server/ParticleGroupData.java
package work.nekow.particledrawing.core.server;

import net.minecraft.world.phys.Vec3;
import work.nekow.particledrawing.api.Color;

import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Holds a group of particles that can be transformed together.
 * The pivot is the reference point for rotations and scaling.
 */
public final class ParticleGroupData {

    private final UUID id;
    private final List<UUID> memberIds;
    private Vec3 pivot;

    ParticleGroupData(UUID id, Vec3 pivot) {
        this.id = id;
        this.pivot = pivot;
        this.memberIds = new CopyOnWriteArrayList<>();
    }

    public UUID id() { return id; }
    public List<UUID> memberIds() { return Collections.unmodifiableList(memberIds); }
    public int size() { return memberIds.size(); }
    public Vec3 pivot() { return pivot; }

    public void setPivot(Vec3 pivot) {
        this.pivot = pivot;
    }

    public void addMember(UUID particleId) {
        if (!memberIds.contains(particleId)) {
            memberIds.add(particleId);
        }
    }

    public void removeMember(UUID particleId) {
        memberIds.remove(particleId);
    }

    public void addMembers(Collection<UUID> ids) {
        for (UUID id : ids) {
            addMember(id);
        }
    }

    public boolean isEmpty() {
        return memberIds.isEmpty();
    }

    public static ParticleGroupData create(UUID id, Vec3 pivot) {
        return new ParticleGroupData(id, pivot);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ParticleGroupData that)) return false;
        return id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    @Override
    public String toString() {
        return "ParticleGroupData{" + id + " members=" + memberIds.size() + "}";
    }
}

File: src/main/kotlin/work/nekow/particledrawing/core/server/ParticleGroupData.kt
package work.nekow.particledrawing.core.server

import net.minecraft.world.phys.Vec3
import java.util.Collections
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList

/**
 * 粒子组数据，管理组成员列表和轴心。
 *
 * @param id 组唯一标识符
 * @param pivot 组轴心坐标
 */
@Suppress("unused")
class ParticleGroupData(
    val id: UUID,
    private var pivot: Vec3
) {

    private val memberIds: MutableList<UUID> = CopyOnWriteArrayList()

    fun memberIds(): List<UUID> = Collections.unmodifiableList(memberIds)
    fun size(): Int = memberIds.size
    fun pivot(): Vec3 = pivot

    fun setPivot(pivot: Vec3) {
        this.pivot = pivot
    }

    fun addMember(particleId: UUID) {
        if (!memberIds.contains(particleId)) {
            memberIds.add(particleId)
        }
    }

    fun removeMember(particleId: UUID) {
        memberIds.remove(particleId)
    }

    fun addMembers(ids: Collection<UUID>) {
        for (id in ids) {
            addMember(id)
        }
    }

    fun isEmpty(): Boolean = memberIds.isEmpty()

    companion object {
        fun create(id: UUID, pivot: Vec3): ParticleGroupData {
            return ParticleGroupData(id, pivot)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ParticleGroupData) return false
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()

    override fun toString(): String = "ParticleGroupData{$id members=${memberIds.size}}"
}
